Fix rounding in cast from TIME to TIMESTAMP

martint opened the pull request:
Fixes #5736

Relates to #5738

martint committed:
Rename newInstance to fromInstant

This is to distinguish between the semantics of newInstance, which is not supposed to do
any conversions (rounding, etc), with the version that constructs a instance from an Instance
by applying any necessary rounding.

martint committed:
Fix rounding in cast from TIME to TIMESTAMP

martint committed:
Make construction of SqlTimestamp and SqlTimestampWithTimeZone stricter

Previously, it was applying rounding rules, which ended up masking potential
problems in operations that are expected to do proper rounding.

martint committed:
Move time->timestamp cast tests to TestTime class

They were in a separate class for when we had to parameterize the tests
for legacy vs new timestamp semantics. That's no longer necessary.

findepi commented:
> Fixes #5736, #5738

To fix https://github.com/prestosql/presto/issues/5738 we should apply checks to other types too:

- length checks for char, varchar
- space padding checks for char
- ...

findepi commented on presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeToTimestampCast.java:
```diff
@@ -56,23 +57,36 @@ public static LongTimestamp castToLong(
+
+    private static long computeEpochMicros(long epochSeconds, long picoFraction)
+    {
+        long microFraction = rescale(picoFraction, 12, 6);
```
Does `rescale` round or truncate?
i checked it truncates, but i'd write this simply as
```
// picoFraction is already rounded to whole micros
long microFraction = picoFraction / PICOSECONDS_PER_MICROSECOND;
```

findepi commented on presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java:
```diff
@@ -43,7 +45,26 @@ public static SqlTimestamp fromMillis(int precision, long millis)
-        return newInstanceWithRounding(precision, epochMicros, picosOfMicro);
+        if (precision <= 6) {
+            if (picosOfMicro != 0) {
+                throw new IllegalArgumentException("Expected picosOfMicro to be 0 for precision <= 6: " + picosOfMicro);
```
Mentioning epochMicros and precision as part of the exc msg may help tracing down the problem, should the exception occur anywhere outside of our tests.

introducing `private void checkArgument(...)` may help increase readability.

findepi commented on presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java:
```diff
@@ -43,7 +45,26 @@ public static SqlTimestamp fromMillis(int precision, long millis)
+                throw new IllegalArgumentException("Expected picosOfMicro to be 0 for precision <= 6: " + picosOfMicro);
+            }
+
+            if (rescale(rescale(epochMicros, 6, precision), precision, 6) != epochMicros) {
```
since you already know precision <= 6, you can more tersely:

```
if (round(epochMicros, 6 - precision) != epochMicros)
```

findepi commented on presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java:
```diff
@@ -43,7 +45,26 @@ public static SqlTimestamp fromMillis(int precision, long millis)
+            }
+        }
+        else {
+            if (rescale(rescale(picosOfMicro, 12, precision), precision, 12) != picosOfMicro) {
```
`round` as above

findepi commented on presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java:
```diff
@@ -43,7 +45,26 @@ public static SqlTimestamp fromMillis(int precision, long millis)
     public static SqlTimestamp newInstance(int precision, long epochMicros, int picosOfMicro)
     {
-        return newInstanceWithRounding(precision, epochMicros, picosOfMicro);
+        if (precision <= 6) {
```
should the `fromMillis` above do rounding or checks?

or maybe we (re)move it, since it is used in tests only

martint replied:
It's supposed to be a convenience method to instantiate a SqlTimestamp of a given precision, including any necessary rounding. The issue with (re)moving it is that it would require making `newInstanceWithRounding` public, and that's not something I want to be visible.

findepi replied:
> The issue with (re)moving it is that it would require making newInstanceWithRounding public, and that's not something I want to be visible.

`fromMillis` currently doesn't call `newInstanceWithRounding`, it calls `newInstance`

(perhaps incorectly so?)

martint replied:
Oh yeah, I fixed that in my branch and forgot to push :)

findepi commented on presto-spi/src/main/java/io/prestosql/spi/type/Timestamps.java:
```diff
@@ -79,10 +79,6 @@ public static long round(long value, int magnitude)
     static long rescale(long value, int fromPrecision, int toPrecision)
     {
-        if (value < 0) {
-            throw new IllegalArgumentException("value must be >= 0");
```
`rescale` method works great when the rounding is known to be not necessary. 
For cases when rounding may be necessary, `round`, `floorDiv` seem to offer well defined semantics.

In this commit you're using the rounding-or-truncation of this method, but i'd change it to expect that value is pre-rounded. (actual check may be too expensive for runtime, idk).

This is more relevant when we allow negative numbers here.

martint replied:
> rescale method works great when the rounding is known to be not necessary.

Correct, and that's the exact purpose of this method. The only reason we had this check is to protect against trying to rescale epoch values which can be negative. But it doesn't matter. I'll get rid of the call to rescale as you suggested above and revert this change.

findepi reviewed (approved):
Couple comments, otherwise LGTM.

Per https://github.com/prestosql/presto/pull/5742#issuecomment-718992077, i think we should not close https://github.com/prestosql/presto/issues/5738 when merging this.

martint committed:
Fix roundDiv when factor is 1

For negative numbers, it will produce a result different
from the original value, which is incorrect.

findepi commented on presto-spi/src/test/java/io/prestosql/spi/type/TestTimestamps.java:
```diff
@@ -0,0 +1,56 @@
+    @Test
+    public void testRoundDiv()
+    {
+        // round towards negative
```
towards negative -> towards zero?
or "round down"

findepi commented on presto-spi/src/test/java/io/prestosql/spi/type/TestTimestamps.java:
```diff
@@ -0,0 +1,56 @@
+        assertThat(roundDiv(-55556, 10000)).isEqualTo(-6);
+        assertThat(roundDiv(-55556, 100000)).isEqualTo(-1);
+
+        // round towards positive
```
"round up"

findepi commented on presto-spi/src/test/java/io/prestosql/spi/type/TestTimestamps.java:
```diff
@@ -0,0 +1,56 @@
+public class TestTimestamps
+{
+    @Test
+    public void testRoundDiv()
```
perhaps add a test with 0 divisor (failure)
it would be great if negative divisor was rejected too

findepi commented on presto-main/src/main/java/io/prestosql/testing/DateTimeTestingUtils.java:
```diff
@@ -63,7 +64,7 @@ public static SqlTimestampWithTimeZone sqlTimestampWithTimeZoneOf(int precision,
 
         long epochMillis = base.toEpochSecond() * MILLISECONDS_PER_SECOND + nanoOfSecond / NANOSECONDS_PER_MILLISECOND;
-        int picosOfMilli = (int) scaleNanosToPicos(nanoOfSecond);
+        int picosOfMilli = (int) (scaleNanosToPicos(nanoOfSecond) % PICOSECONDS_PER_MILLISECOND);
```
I dont understand the % here. If the value can `>= PICOSECONDS_PER_MILLISECOND`,  it seems we're discarding some information without compensating it (with eg `epochMillis++` or whatever is appropriate)

martint replied:
The new code is actually correct. It works because of the line above. `epochMillis` is computed by taking into account `nanoOfSecond / NANOSECONDS_PER_MILLISECOND`. The remainder is stored in `picosOfMilli`.

It would actually be clearer to do it the other way around. I'll update it.

```java
long epochMillis = base.toEpochSecond() * MILLISECONDS_PER_SECOND + nanoOfSecond / NANOSECONDS_PER_MILLISECOND;
int picosOfMilli = (nanoOfSecond % NANOSECONDS_PER_MILLISECOND) * PICOSECONDS_PER_NANOSECOND;
```

findepi replied:
yeah, that's so clear that even i can understand. thanks

findepi commented on presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java:
```diff
@@ -38,12 +39,24 @@
+            checkArgument(round(picosOfMicro, 12 - precision) == picosOfMicro, "Expected 0s for digits beyond precision %s: picosOfMicro = %s", precision, picosOfMicro);
+        }
+
+        if (picosOfMicro < 0 || picosOfMicro > PICOSECONDS_PER_MICROSECOND) {
+            throw new IllegalArgumentException("picosOfMicro is out of range: " + picosOfMicro);
+        }
```
This could use `checkArgument` too

findepi replied:
Same reasoning either applies or doesn't apply to other checks above, right?

martint replied:
Yes, you're right. For some reason, I was thinking that they were in a conditional that was only evaluated sometimes, but that's obviously not the case. I'll revert the changes

martint replied:
Yes, you're right. For some reason, I was thinking that they were in a conditional that was only evaluated sometimes, but that's obviously not the case. I'll revert the changes

findepi commented on presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestampWithTimeZone.java:
```diff
@@ -47,7 +48,19 @@ public static SqlTimestampWithTimeZone fromInstant(int precision, Instant instan
+            checkArgument(round(picosOfMilli, 12 - precision) == picosOfMilli, "Expected 0s for digits beyond precision %s: picosOfMilli = %s", precision, picosOfMilli);
+        }
+
+        if (picosOfMilli < 0 || picosOfMilli > PICOSECONDS_PER_MILLISECOND) {
+            throw new IllegalArgumentException("picosOfMilli is out of range: " + picosOfMilli);
+        }
```
This could use `checkArgument` too

findepi commented on presto-spi/src/test/java/io/prestosql/spi/type/TestSqlTimestamp.java:
```diff
@@ -40,78 +40,6 @@ public void testBaseline()
-        assertThat(newInstance(3, 111111, 111111).toString()).isEqualTo("1970-01-01 00:00:00.111");
-        assertThat(newInstance(4, 111111, 111111).toString()).isEqualTo("1970-01-01 00:00:00.1111");
-        assertThat(newInstance(5, 111111, 111111).toString()).isEqualTo("1970-01-01 00:00:00.11111");
-        assertThat(newInstance(6, 111111, 111111).toString()).isEqualTo("1970-01-01 00:00:00.111111");
```
This deletion is part `Make construction of SqlTimestamp and SqlTimestampWithTimeZone stricter` commit.
Would it mnake sense toi keep the test, replacing `newInstance` with `newInstanceWithRounding` here?

findepi replied:
you can make `newInstanceWithRounding` package private, visiblefortesting

anyway, up to you

martint committed:
Add tests for SqlTimestamp.fromMillis

martint committed:
Inline method
